textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/GramBooleanQuery.java

```diff
@@ -293,6 +294,133 @@ private static String toLuceneQueryString(GramBooleanQuery query) {
 			}
 		}
 	}
+	
+	public boolean isEmpty() {
+		if (this.operandSet.size() > 0) {
+			return false;
+		}
+		for (GramBooleanQuery subQuery : this.subQuerySet) {
+			if (! subQuery.isEmpty()) {
+				return false;
+			}
+		}
+		return true;
+	}
+	
+	// "AND" two DNF tree
+	// Apply distributive laws: 
+	// (a OR b) AND (c OR d) --> (a AND c) OR (a AND d) OR (b AND c) OR (c AND d)
```

Comment:
So we are not changing from Conjunctive form, we are changing from a very messy form, with possibly lots of redundant information. I'll give you an example.

Regex is `abc?pqr?`, which can possibly match: abpq, abpqr, abcpq, abcpqr

The original query generated by our program is `(((bpq AND abp) OR (abc AND cpq AND bcp)) AND (bpq OR cpq) AND (bpq OR cpq OR (pqr AND cpq) OR (pqr AND bpq)) AND (pqr OR bpq OR cpq) AND (abc OR abp))`. It is so long and complicated that we can only print it out, go through it manually and see if it's correct (it is correct BTW).

When the regex goes a little bit complicated, the query generated is like crazy and we can't write automated test on it. So professor Li suggested us to translate it to DNF, (I believe CNF, Conjunctive normal form won't make a difference). 

So this is the code to transform the query to DNF, and then I can easily do some simplifications on it. (in `simplifyDNF` function).

Here's the query after transforming to DNF and applying simplification:
`(abp AND bpq) OR (abc AND bcp AND cpq)`

Remember the regex can match abpq, abpqr, abcpq, abcpqr, 
abpqr is redundant because of abpq, abcpqr is redundant because of abcpq.

We now get a much better query :)